PartitionParmetis.h

```diff
@@ -0,0 +1,105 @@
+/**
+ * @file
+ *  This file is part of PUML
+ *
+ *  For conditions of distribution and use, please see the copyright
+ *  notice in the file 'COPYING' at the root directory of this package
+ *  and the copyright notice at https://github.com/TUM-I5/PUMGen
+ *
+ * @copyright 2019 Technische Universitaet Muenchen
+ * @author Sebastian Rettenberger <sebastian.rettenberger@tum.de>
+ * @author David Schneller <david.schneller@tum.de>
+ */
+
+#ifndef PUML_PARTITIONPARMETIS_H
+#define PUML_PARTITIONPARMETIS_H
+
+#ifdef USE_MPI
+#include <mpi.h>
+#endif // USE_MPI
+
+#ifndef USE_PARMETIS
+#warning ParMETIS is not enabled.
+#endif
+
+#include <metis.h>
+#include <parmetis.h>
+
+#include <cassert>
+
+#include "PartitionBase.h"
+#include "PartitionGraph.h"
+#include "Topology.h"
+
+namespace PUML
+{
+
+template<TopoType Topo>
+class PartitionParmetis : public PartitionBase<Topo>
+{
+
+public:
+	PartitionParmetis(int mode) :mode(mode) {
+
+	}
+#ifdef USE_MPI
+	virtual PartitioningResult partition(int* partition, const PartitionGraph<Topo>& graph, const PartitionTarget& target, int seed = 1)
+	{
+		auto comm = graph.comm();
+		std::vector<idx_t> vtxdist(graph.vertexDistribution().begin(), graph.vertexDistribution().end());
+		std::vector<idx_t> xadj(graph.adjDisp().begin(), graph.adjDisp().end());
+		std::vector<idx_t> adjncy(graph.adj().begin(), graph.adj().end());
+		std::vector<idx_t> vwgt(graph.vertexWeights().begin(), graph.vertexWeights().end());
+		std::vector<idx_t> adjwgt(graph.edgeWeights().begin(), graph.edgeWeights().end());
+		auto cell_count = graph.localVertexCount();
+
+		idx_t ncon = graph.vertexWeightCount();
+		if (ncon == 0) ncon = 1;
+		idx_t nparts = target.vertexCount();
+		std::vector<real_t> tpwgts(nparts * ncon, static_cast<real_t>(1.) / nparts);
+		if (!target.vertexWeightsUniform()) {
+			for (idx_t i = 0; i < target.vertexCount(); i++) {
+				for (idx_t j = 0; j < ncon; ++j) {
+					tpwgts[i*ncon + j] = target.vertexWeights()[i];
+				}
+			}
+		}
```

Comment:
That's a good proposal, but there are two things to consider:
1. The current main branch also broadcasts target weights, so this PR does not remove functionality.
2. SCOTCH and ParHIP do not support multiple vertex weights (@davschneller correct me if I'm wrong).
Therefore, I suggest to realize multiple vertex weights in a different PR, where one can think about what to do when the graph partitioning library does not support multiple weights. (One should also think about whether one really needs that feature if there are few graph partitioning libraries that support it.)

Reply:
Correct, they do not; at least to my knowledge (I'll have to double-check for PT-SCOTCH anyways).

While I'd really like to agree and implement it right now, I'd also be rather for putting it into a future PR—mainly since I kept this one here is already out for a pretty long time, so I guess it would be good for it to be merged rather soon.

A future PR could then also handle better topologies (i.e. `PartitionTarget`) handling for e.g. PT-SCOTCH, or implement other partitioners—or implement meta-heuristics which handle the topology or some other aspects (like multiple weights, maybe?).

Reply:
Hi, yes. I agree. Let's leave it for a future PR